# Домашнє завдання до теми "Основи роботи з даними у Pandas"

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import numpy as np
import pandas as pd

In [5]:
data_path = r"C:\Users\Вальчук Богдан\OneDrive\Документы\Курс з дата аналітики\Модуль 10. Бібліотека Pandas ключові концепції та методи роботи з даними\Data\supermarket_sales.csv"
df = pd.read_csv(data_path)

1. За допомогою бібліотеки Pandas виведіть перші 4 і останні 4 записи у наборі даних (це дві окремі команди). Треба саме 4! Не 5.

In [8]:
df.head(4)


,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
0,750-67-8428,A,Yangon,Member,Female,Health and beauty,74.69,7,26.1415,548.9715,1/5/2019,13:08,Ewallet,522.83,4.761905,26.1415,9.1
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.8200,80.2200,3/8/2019,10:29,Cash,76.40,4.761905,3.8200,9.6
2,631-41-3108,A,Yangon,Normal,Male,Home and lifestyle,46.33,7,16.2155,340.5255,3/3/2019,13:23,Credit card,324.31,4.761905,16.2155,7.4
3,123-19-1176,A,Yangon,Member,Male,Health and beauty,58.22,8,23.2880,489.0480,1/27/2019,20:33,Ewallet,465.76,4.761905,23.2880,8.4


In [9]:
df.tail(4)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
996,303-96-2227,B,Mandalay,Normal,Female,Home and lifestyle,97.38,10,48.690,1022.490,3/2/2019,17:16,Ewallet,973.80,4.761905,48.690,4.4
997,727-02-1313,A,Yangon,Member,Male,Food and beverages,31.84,1,1.592,33.432,2/9/2019,13:22,Cash,31.84,4.761905,1.592,7.7
998,347-56-2442,A,Yangon,Normal,Male,Home and lifestyle,65.82,1,3.291,69.111,2/22/2019,15:33,Cash,65.82,4.761905,3.291,4.1
999,849-09-3807,A,Yangon,Member,Female,Fashion accessories,88.34,7,30.919,649.299,2/18/2019,13:28,Cash,618.38,4.761905,30.919,6.6


Очікуваний результат:

![](Image%201.png)




2. Виведіть рядки даних з індексами з 615 по 619 включно.

In [11]:
df[615:620]

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
615,509-29-3912,A,Yangon,Member,Female,Sports and travel,38.72,9,17.4240,365.9040,3/20/2019,12:24,Ewallet,348.48,4.761905,17.4240,4.2
616,327-40-9673,B,Mandalay,Member,Male,Sports and travel,72.60,6,21.7800,457.3800,1/13/2019,19:51,Cash,435.60,4.761905,21.7800,6.9
617,840-19-2096,C,Naypyitaw,Member,Male,Electronic accessories,87.91,5,21.9775,461.5275,3/14/2019,18:10,Ewallet,439.55,4.761905,21.9775,4.4
618,828-46-6863,A,Yangon,Member,Male,Food and beverages,98.53,6,29.5590,620.7390,1/23/2019,11:22,Credit card,591.18,4.761905,29.5590,4.0
619,641-96-3695,C,Naypyitaw,Member,Female,Fashion accessories,43.46,6,13.0380,273.7980,2/7/2019,17:55,Ewallet,260.76,4.761905,13.0380,8.5


Очікуваний результат:

![](Image2.png)

3. Виведіть назви колонок у наборі даних разом з їхніми типами. Це можна зробити кількома способами, важливо аби в результаті були назви колонок і відповідні їм типи.

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 17 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Invoice ID               1000 non-null   object 
 1   Branch                   1000 non-null   object 
 2   City                     1000 non-null   object 
 3   Customer type            1000 non-null   object 
 4   Gender                   1000 non-null   object 
 5   Product line             1000 non-null   object 
 6   Unit price               1000 non-null   float64
 7   Quantity                 1000 non-null   int64  
 8   Tax 5%                   1000 non-null   float64
 9   Total                    1000 non-null   float64
 10  Date                     1000 non-null   object 
 11  Time                     1000 non-null   object 
 12  Payment                  1000 non-null   object 
 13  cogs                     1000 non-null   float64
 14  gross margin percentage  

Очікуваний результат (один з можливих форматів):

![](Image3.png)

4. Користуючись методом `DataFrame.describe()` виявіть 1, 25, 50, 75 і 99 персентилі колонки `gross income`. Знайдіть, яке значення відповідає 75отму персентилю цієї колонки і що це означає? Як би Ви його пояснили простими словами колегам? Відповідь напишіть текстом в окремій клітинці типу Markdown.

In [18]:
df.describe()

,Unit price,Quantity,Tax 5%,Total,cogs,gross margin percentage,gross income,Rating
count,1000.000000,1000.000000,1000.000000,1000.000000,1000.00000,1.000000e+03,1000.000000,1000.00000
mean,55.672130,5.510000,15.379369,322.966749,307.58738,4.761905e+00,15.379369,6.97270
std,26.494628,2.923431,11.708825,245.885335,234.17651,6.131498e-14,11.708825,1.71858
min,10.080000,1.000000,0.508500,10.678500,10.17000,4.761905e+00,0.508500,4.00000
25%,32.875000,3.000000,5.924875,124.422375,118.49750,4.761905e+00,5.924875,5.50000
50%,55.230000,5.000000,12.088000,253.848000,241.76000,4.761905e+00,12.088000,7.00000
75%,77.935000,8.000000,22.445250,471.350250,448.90500,4.761905e+00,22.445250,8.50000
max,99.960000,10.000000,49.650000,1042.650000,993.00000,4.761905e+00,49.650000,10.00000


In [20]:
df['gross income'].describe(percentiles=[0.01, 0.25, 0.5, 0.75, 0.99])

count    1000.000000
mean       15.379369
std        11.708825
min         0.508500
1%          0.913735
25%         5.924875
50%        12.088000
75%        22.445250
99%        45.250750
max        49.650000
Name: gross income, dtype: float64

75% - 22.445250  Це означає, що 75% клієнтів мають показник gross income нижче 22,45, а 25% клієнтів - вищий

Очікуваний результат:

![](Image4.png)

5. Виведіть кількість інвойсів за кожним унікальним містом в наборі даних.

  Підказка: інвойс - це один рядок в цьому наборі даних, ми можемо цим скористатись аби швидко отримати відповідь.

In [31]:
df["City"].value_counts() 

City
Yangon       340
Mandalay     332
Naypyitaw    328
Name: count, dtype: int64

Очікуваний результат:

![](Image5.png)

6. Виведіть, який відсоток інвойсів припадає на кожне унікальне місто.

In [32]:
df["City"].value_counts(normalize=True) 

City
Yangon       0.340
Mandalay     0.332
Naypyitaw    0.328
Name: proportion, dtype: float64

Очікуваний результат:

![](Image6.png)

7.1. Використовуючи фільтрацію даних, знайдіть середнє значення `gross income` для міст `Yangon` та `Mandalay`. В якому місті воно більше?

7.2 Сформулюйте 3 гіпотези, чому в одному місті gross income вище, як можна дізнатись чи правдива ця гіпотеза з використанням даних (наявних і додаткових), та як висновки з цих гіпотез можна використати для прийняття рішень як підвищити прибуток в мережі супермаркетів. Відповідь напишіть текстом в окремій клітинці типу Markdown.

Це завдання - на продуктове мислення. Якщо не знаєте, які гіпотези сформулювати - напишіть ті, які здаються вам зовсім примітивними. Важливо навчитись висловлювати свої гіпотези щодо даних, і з часом та досвідом вони будуть ставати щоразу глибше і точніше. Весь аналіз даних будується на тому, що ми перевіряємо певні гіпотези та робимо висновки і рекоемндації на основі своїх досліджень.

In [43]:
df.groupby('City')['gross income'].mean()

City
Mandalay     15.232024
Naypyitaw    16.052367
Yangon       14.874001
Name: gross income, dtype: float64

Очікуваний результат:

![](Image7.png)

Yangon - 14.87 Mandalay - 15.23 gross income більший у місті Mandalay. Перед тим як сформувати гіпотези, потрібно сказати, що gross income тут математично дорівнює виручці мінус собівартість, але фактично це просто сума податку, а не реальний торговий прибуток (маржа продавця). Тобто в цьому конкретному датасеті "gross income" — оманлива назва: це не "скільки заробив магазин понад собівартість" у звичному бізнес-сенсі (продажна ціна мінус закупівельна), а буквально сума податку з транзакції. В масиві даних це вираховуєть з цього виразу Total = cogs + Tax 5%. Тобто gross margin percentage = 4.761905% — це фіксована податкова ставка (5/105×100), а не змінна бізнес-маржа. Тож виставляти гіпотези відносно впливу на собівартість сенсу немає.

Нижче перевіряла гіпотезу чи впливає тип клієнта на кількість покупок. Не впливає.

In [38]:
df.groupby(['City', 'Customer type'])['Quantity'].mean()
df.groupby('City')['Customer type'].value_counts(normalize=True)

City       Customer type
Mandalay   Normal           0.503012
           Member           0.496988
Naypyitaw  Member           0.515244
           Normal           0.484756
Yangon     Normal           0.508824
           Member           0.491176
Name: proportion, dtype: float64

In [39]:
df.groupby('City')['Quantity'].mean()

City
Mandalay     5.481928
Naypyitaw    5.582317
Yangon       5.467647
Name: Quantity, dtype: float64

## Перша гіпотеза: 
- може впливати ціна товару, що продається. Можемо перевірити через наші дані.
## Друга гіпотеза: 
- які саме товари(асортимент) продається. Можемо перевірити через наші дані.
## Третя гіпотеза: 
- соціально-економічні характеристики міста (чисельність населення, рівень розвитку інфраструктури й промисловості, купівельна спроможність жителів, туристична привабливість).
Цю гіпотезу можна перевірити через:
- Офіційна статистика населення міст М'янми (перепис/держстат)
- Регіональний ВВП на душу населення чи середня зарплата по регіонах
- Дані про туристичний потік (кількість відвідувачів визначних місць по містах/рік)
- Порівняти ці зовнішні показники з середнім Total/gross income по місту — чи є кореляція

Перевіряємо середню ціну по Product line відносно кожного міста

In [51]:
df[df['City'].isin(['Mandalay', 'Yangon'])].groupby(['City', 'Product line'])['Unit price'].mean()

City      Product line          
Mandalay  Electronic accessories    49.854182
          Fashion accessories       54.843871
          Food and beverages        55.540000
          Health and beauty         58.185660
          Home and lifestyle        55.514000
          Sports and travel         59.678065
Yangon    Electronic accessories    54.871167
          Fashion accessories       56.670392
          Food and beverages        54.974483
          Health and beauty         49.862340
          Home and lifestyle        55.845692
          Sports and travel         55.610339
Name: Unit price, dtype: float64

Перевіряємо частку Product line у кожному місті

In [52]:
df[df['City'].isin(['Mandalay', 'Yangon'])].groupby('City')['Product line'].value_counts(normalize=True)

City      Product line          
Mandalay  Fashion accessories       0.186747
          Sports and travel         0.186747
          Electronic accessories    0.165663
          Health and beauty         0.159639
          Food and beverages        0.150602
          Home and lifestyle        0.150602
Yangon    Home and lifestyle        0.191176
          Electronic accessories    0.176471
          Sports and travel         0.173529
          Food and beverages        0.170588
          Fashion accessories       0.150000
          Health and beauty         0.138235
Name: proportion, dtype: float64

Тепер видно розкид: у Mandalay Sports and travel одночасно і дорога, і з високою часткою — це підсилює її внесок у gross income. У Yangon Home and lifestyle — теж висока частка, але ціна середня, тоді як Health and beauty там дешевша і з найменшою часткою.

# Як підвищити прибуток в мережі супермаркетів
# З гіпотези 1 (ціна товару):
Маркетинг/мерчандайзинг можуть стимулювати продажі саме дорожчих категорій (Sports and travel, Health and beauty) — через розміщення на видному місці, бандли, персоналізовані рекомендації касирам/додатку.
# З гіпотези 2 (мікс асортименту):
Якщо в конкретному місті продається забагато дешевих категорій — це сигнал переглянути асортиментну політику саме там: чи достатньо представлені дорожчі лінійки, чи потрібна локальна промо-кампанія на них.
# З гіпотези 2 (мікс асортименту):
Якщо підтвердиться зв'язок з туристичним потоком чи доходами населення — має сенс адаптувати асортимент і цінову політику під конкретне місто (наприклад, туристам пропонувати преміум-категорії), а не застосовувати однакову стратегію по всій мережі.

8. Для продовження дослідження різниці у прибутках в магазинах давайте зʼясуємо, чи відрізняється маржа ('gross margin percentage') для різних інвойсів в магазині? Перевірте це програмно на даних виходячи з того функціоналу, який ми вже пройшли у лекціях.

Дайте на основі отриманого результату відповідь: "У всіх всіх інвойсах однакова маржа" або "Маржа відрізняється для різних інвойсів".

In [54]:
df['gross margin percentage']

0      4.761905
1      4.761905
2      4.761905
3      4.761905
4      4.761905
         ...   
995    4.761905
996    4.761905
997    4.761905
998    4.761905
999    4.761905
Name: gross margin percentage, Length: 1000, dtype: float64

Очікуваний результат: відповідь щодо маржі в усіх інвойсах на основі отриманих результатів

9. Врахуовуючи висновок в попередньому завданні, дослідимо як продаються в цих містах різні категорії товарів. Виведіть відсотковий розпоіділ кількості інвойсів по категоріям товарів `Product line` в містах `Yangon` та `Mandalay`.

In [53]:
df[df['City'].isin(['Mandalay', 'Yangon'])].groupby('City')['Product line'].value_counts(normalize=True)

City      Product line          
Mandalay  Fashion accessories       0.186747
          Sports and travel         0.186747
          Electronic accessories    0.165663
          Health and beauty         0.159639
          Food and beverages        0.150602
          Home and lifestyle        0.150602
Yangon    Home and lifestyle        0.191176
          Electronic accessories    0.176471
          Sports and travel         0.173529
          Food and beverages        0.170588
          Fashion accessories       0.150000
          Health and beauty         0.138235
Name: proportion, dtype: float64

Очікуваний результат для кожного з міст:

![](Image8.png)

![](Image9.png)